src/Compilers/Core/Portable/Compilation/CompilationOptions.cs

```diff
@@ -466,6 +471,47 @@ public CompilationOptions WithStrongNameProvider(StrongNameProvider provider)
         /// </summary>
         internal abstract void ValidateOptions(ArrayBuilder<Diagnostic> builder);
 
+        internal void ValidateOptions(ArrayBuilder<Diagnostic> builder, CommonMessageProvider messageProvider)
+        {
+            if (!CryptoPublicKey.IsEmpty)
+            {
+                if (CryptoKeyFile != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
+                        Location.None, nameof(CryptoPublicKey), nameof(CryptoKeyFile)));
+                }
+
+                if (CryptoKeyContainer != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
+                        Location.None, nameof(CryptoPublicKey), nameof(CryptoKeyContainer)));
+                }
+            }
+
+            if (PublicSign)
+            {
+                if (CryptoKeyFile != null && !PathUtilities.IsAbsolute(CryptoKeyFile))
+                {
+                    // TODO(https://github.com/dotnet/roslyn/issues/9153):
+                    // Produce better diagnostic message for passing a key file with a relative path
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_BadCompilationOptionValue,
+                        Location.None, nameof(CryptoKeyFile), CryptoKeyFile));
+                }
+
+                if (CryptoKeyContainer != null)
+                {
+                    builder.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_MutuallyExclusiveOptions,
```

Comment:
We did not report this error, but compilation fails on non-Windows platforms with `/keycontainer` set since it 1) is Windows-only and 2) relies on the DesktopStrongNameProvider.

This would not be considered a breaking change because we have never shipped a stable build with `/publicsign` available.